### PR TITLE
Base predictor implementation: NN, LinReg, RF

### PR DESCRIPTION
**Pull request by danyoungday**
Fixed a minor merge issue in data file.
Fixed an issue where the Evaluator got removed somehow (oops, not sure how this happened)

Implemented abstract Predictor class into NeuralNetPredictor
Implemented abstract Predictor class into abstract SKLearnPredictor class which then is implemented as LinearRegressionPredictor and RandomForestPredictors. I think this is a quite elegant way to handle using SKLearn predictors.

### CODE REVIEW

File: use_cases/eluc/data/eluc_data.py [hunk 1]
```diff
@@ -138,21 +138,6 @@ def hf_to_df(self, hf_repo):
         df = df.set_index(["time", "lat", "lon"], drop=False)
         return df
 
-
```
Comment (danyoungday) [inline use_cases/eluc/data/eluc_data.py:141]: Made a small mistake with the merge resolver. I don't think this actually breaks anything though thankfully

File: use_cases/eluc/predictors/neural_network/neural_net_predictor.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        self.y = torch.tensor(y)
+
+    def __len__(self):
+        return len(self.X)
+
+    def __getitem__(self, idx: int) -> tuple:
+        return self.X[idx], self.y[idx]
+
+
+class ELUCNeuralNet(torch.nn.Module):
```
Comment (danyoungday) [inline use_cases/eluc/predictors/neural_network/neural_net_predictor.py:38]: Simple PyTorch neural network implementation

File: use_cases/eluc/predictors/neural_network/neural_net_predictor.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    """
+    Custom torch neural network module.
+    :param in_size: number of input features
+    :param hidden_sizes: list of hidden layer sizes
+    :param linear_skip: whether to concatenate input to hidden layer output
+    :param dropout: dropout probability
+    """
+    class EncBlock(torch.nn.Module):
```
Comment (danyoungday) [inline use_cases/eluc/predictors/neural_network/neural_net_predictor.py:46]: EncBlock class allows us to easily give the NN more layers

File: use_cases/eluc/predictors/neural_network/neural_net_predictor.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    Has the option to use wide and deep, concatenating the input to the output of the hidden layers
+    in order to take advantage of the linear relationship in the data.
+    Data is automatically standardized and the scaler is saved with the model.
+    """
+    def __init__(self, features=None, label=None, hidden_sizes=[4096], linear_skip=True, dropout=0, device="mps",
+            epochs=3, batch_size=2048, optim_params={}, train_pct=1, step_lr_params={"step_size": 1, "gamma": 0.1}): 
+        # Model setup params
+        self.model = None
+        self.features = features
+        self.label = label
```
Comment (danyoungday) [inline use_cases/eluc/predictors/neural_network/neural_net_predictor.py:98]: We make sure to keep track of the features and label for when we load the model later

File: use_cases/eluc/predictors/neural_network/neural_net_predictor.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        :param y_val: validation labels.
+        :param X_test: test data, may be unscaled and have excess features.
+        :param y_test: test labels.
+        :param log_path: path to log training data to tensorboard.
+        :param verbose: whether to print progress bars.
+        :return: dictionary of results from training containing time taken, best epoch, best loss, and test loss if applicable.
+        """
+        if not self.features:
+            self.features = X_train.columns.tolist()
+        self.label = y_train.name
```
Comment (danyoungday) [inline use_cases/eluc/predictors/neural_network/neural_net_predictor.py:182]: Features and label are set here during fitting. If we haven't specified a set of features to use we just take them all from X_train. The label is taken from y_train.

File: use_cases/eluc/predictors/neural_network/neural_net_predictor.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+            self.model.eval()
+            for X, _ in test_dl:
+                X = X.to(self.device)
+                pred_list.append(self.model(X))
+
+        if len(pred_list) > 1:
+            y_pred = torch.concatenate(pred_list, dim=0).cpu().numpy()
+        else:
+            y_pred = pred_list[0].cpu().numpy()
+        return pd.DataFrame(y_pred, index=X_test.index, columns=[self.label])
```
Comment (danyoungday) [inline use_cases/eluc/predictors/neural_network/neural_net_predictor.py:297]: Note: I also copy the index from X_test. I don't believe we do this in ESP because I was having some issues with it. In the ELUC data case where we have a more complicated index (time, lat, lon) rather than just numerically indexed,  this saves the trouble later of having to manually reattach the indices.

File: use_cases/eluc/predictors/predictor_evaluator.py
Comment (danyoungday) [inline use_cases/eluc/predictors/predictor_evaluator.py:1]: Not sure how this got removed but it's back

File: use_cases/eluc/predictors/sklearn/sklearn_predictor.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+from abc import ABC
+
+import joblib
+import pandas as pd
+from sklearn.linear_model import LinearRegression
+from sklearn.ensemble import RandomForestRegressor
+
+from predictors.predictor import Predictor
+
+class SKLearnPredictor(Predictor, ABC):
```
Comment (danyoungday) [inline use_cases/eluc/predictors/sklearn/sklearn_predictor.py:13]: The SKLearnPredictor pretty much handles everything since the SKLearn models all inherit from something behind the scenes that lets us do this.

File: use_cases/eluc/predictors/sklearn/sklearn_predictor.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        return pd.DataFrame(y_pred, index=X_test.index, columns=[self.label])
+    
+class LinearRegressionPredictor(SKLearnPredictor):
+    """
+    Simple linear regression predictor.
+    See SKLearnPredictor for more details.
+    """
+    def __init__(self, features=None, **kwargs):
+        super().__init__(features)
+        self.model = LinearRegression(**kwargs)
```
Comment (danyoungday) [inline use_cases/eluc/predictors/sklearn/sklearn_predictor.py:83]: All we need to do to implement LinearRegression is specify what the model is in the constructor. The abstract class handles the rest for us! Nice and clean and easy!

File: use_cases/eluc/predictors/predictor_evaluator.py
Comment (ofrancon) [reply use_cases/eluc/predictors/predictor_evaluator.py:1]: Maybe got deleted when you renamed it

File: use_cases/eluc/predictors/sklearn/sklearn_predictor.py (hunk 8 above)
Comment (ofrancon) [reply use_cases/eluc/predictors/sklearn/sklearn_predictor.py:83]: Looks good indeed. I'm not sure how the `label` is passed down to the super class though. Should you have a `label` param along with the `features` one in the constructor here?

**Review by ofrancon — APPROVED** [review]
lgtm
